File: backend/src/services/urlService.ts
import prisma from '../config/db';
import bcrypt from 'bcryptjs';
import { generateShortCode } from '../utils/helpers';
import ErrorResponse from '../utils/errorResponse';
export interface CreateUrlParams {
  original_url: string;
  custom_alias?: string | null;
  expires_at?: string | null;
  user_id?: string | null;
  password?: string | null;
  is_one_time?: boolean;
}
export interface GetUrlsParams {
  user_id: string;
  search?: string;
  isActive?: boolean;
  page?: number;
  limit?: number;
}
export interface UpdateUrlParams {
  id: string;
  user_id: string;
  original_url?: string;
  custom_alias?: string | null;
  expires_at?: string | null;
  is_active?: boolean;
  password?: string | null;
  is_one_time?: boolean;
}
export const createShortUrl = async ({ original_url, custom_alias, expires_at, user_id, password, is_one_time }: CreateUrlParams) => {
  let shortCode = '';
  if (!user_id) {
    custom_alias = null;
    const expiry = new Date();
    expiry.setDate(expiry.getDate() + 7);
    expires_at = expiry.toISOString();
  }
  if (custom_alias) {
    const existing = await prisma.url.findFirst({
      where: {
        OR: [
          { short_code: custom_alias },
          { custom_alias: custom_alias }
        ]
      }
    });
    if (existing) {
      throw new ErrorResponse('This custom alias is already taken.', 400);
    }
    shortCode = custom_alias;
  } else {
    let isUnique = false;
    let attempts = 0;
    while (!isUnique && attempts < 10) {
      shortCode = generateShortCode(6);
      const existing = await prisma.url.findFirst({
        where: {
          OR: [
            { short_code: shortCode },
            { custom_alias: shortCode }
          ]
        }
      });
      if (!existing) {
        isUnique = true;
      }
      attempts++;
    }
    if (!isUnique) {
      throw new ErrorResponse('Failed to generate a unique short URL code. Please try again.', 500);
    }
  }
  const expirationDate = expires_at ? new Date(expires_at) : null;
  let passwordHash = null;
  if (password) {
    const salt = await bcrypt.genSalt(10);
    passwordHash = await bcrypt.hash(password, salt);
  }
  return await prisma.url.create({
    data: {
      original_url,
      short_code: shortCode,
      custom_alias: custom_alias || null,
      expires_at: expirationDate,
      password_hash: passwordHash,
      is_one_time: is_one_time || false,
      user_id: user_id || null
    }
  });
};
export const getUrls = async ({ user_id, search = '', isActive, page = 1, limit = 10 }: GetUrlsParams) => {
  const skip = (page - 1) * limit;
  const where: any = {
    user_id,
    AND: []
  };
  if (search) {
    where.AND.push({
      OR: [
        { original_url: { contains: search, mode: 'insensitive' } },
        { short_code: { contains: search, mode: 'insensitive' } },
        { custom_alias: { contains: search, mode: 'insensitive' } }
      ]
    });
  }
  if (isActive !== undefined) {
    where.AND.push({ is_active: isActive });
  }
  if (where.AND.length === 0) {
    delete where.AND;
  }
  const [urls, total] = await prisma.$transaction([
    prisma.url.findMany({
      where,
      orderBy: { created_at: 'desc' },
      skip,
      take: limit
    }),
    prisma.url.count({ where })
  ]);
  return {
    urls,
    pagination: {
      total,
      page,
      limit,
      pages: Math.ceil(total / limit)
    }
  };
};
export const updateUrl = async ({ id, user_id, original_url, custom_alias, expires_at, is_active, password, is_one_time }: UpdateUrlParams) => {
  const url = await prisma.url.findUnique({
    where: { id }
  });
  if (!url) {
    throw new ErrorResponse('URL not found.', 404);
  }
  if (url.user_id !== user_id) {
    throw new ErrorResponse('Not authorized to edit this URL.', 403);
  }
  const updateData: any = {};
  if (original_url !== undefined) {
    updateData.original_url = original_url;
  }
  if (is_active !== undefined) {
    updateData.is_active = is_active;
  }
  if (is_one_time !== undefined) {
    updateData.is_one_time = is_one_time;
  }
  if (expires_at !== undefined) {
    updateData.expires_at = expires_at ? new Date(expires_at) : null;
  }
  if (password !== undefined) {
    if (password) {
      const salt = await bcrypt.genSalt(10);
      updateData.password_hash = await bcrypt.hash(password, salt);
    } else {
      updateData.password_hash = null;
    }
  }
  if (custom_alias !== undefined && custom_alias !== url.custom_alias) {
    if (custom_alias) {
      const existing = await prisma.url.findFirst({
        where: {
          id: { not: id },
          OR: [
            { short_code: custom_alias },
            { custom_alias: custom_alias }
          ]
        }
      });
      if (existing) {
        throw new ErrorResponse('This custom alias is already taken.', 400);
      }
      updateData.custom_alias = custom_alias;
      updateData.short_code = custom_alias;
    } else {
      let shortCode = '';
      let isUnique = false;
      let attempts = 0;
      while (!isUnique && attempts < 10) {
        shortCode = generateShortCode(6);
        const existing = await prisma.url.findFirst({
          where: {
            id: { not: id },
            OR: [
              { short_code: shortCode },
              { custom_alias: shortCode }
            ]
          }
        });
        if (!existing) {
          isUnique = true;
        }
        attempts++;
      }
      if (!isUnique) {
        throw new ErrorResponse('Failed to generate a unique short URL code. Please try again.', 500);
      }
      updateData.custom_alias = null;
      updateData.short_code = shortCode;
    }
  }
  return await prisma.url.update({
    where: { id },
    data: updateData
  });
};
export const deleteUrl = async (id: string, user_id: string) => {
  const url = await prisma.url.findUnique({
    where: { id }
  });
  if (!url) {
    throw new ErrorResponse('URL not found.', 404);
  }
  if (url.user_id !== user_id) {
    throw new ErrorResponse('Not authorized to delete this URL.', 403);
  }
  await prisma.url.delete({
    where: { id }
  });
  return { success: true };
};
export const resolveUrlAndTrackVisit = async (shortCode: string, ip: string | undefined, userAgent: string | undefined, password?: string) => {
  const url = await prisma.url.findFirst({
    where: {
      OR: [
        { short_code: shortCode },
        { custom_alias: shortCode }
      ]
    }
  });
  if (!url) {
    throw new ErrorResponse('Short URL not found.', 404);
  }
  if (!url.is_active) {
    throw new ErrorResponse('This URL is currently inactive.', 400);
  }
  if (url.expires_at && new Date(url.expires_at) < new Date()) {
    throw new ErrorResponse('This short URL has expired.', 410);
  }
  console.log(`Resolving URL: ${url.short_code}, password_hash type: ${typeof url.password_hash}, value:`, url.password_hash);
  if (url.password_hash !== null && url.password_hash !== '' && url.password_hash.trim().length > 0) {
    if (!password) {
      throw new ErrorResponse('PASSWORD_REQUIRED', 403);
    }
    const isMatch = await bcrypt.compare(password, url.password_hash);
    if (!isMatch) {
      throw new ErrorResponse('Invalid password.', 403);
    }
  }
  const transactionRequests: any[] = [
    prisma.url.update({
      where: { id: url.id },
      data: { click_count: { increment: 1 } }
    }),
    prisma.visit.create({
      data: {
        url_id: url.id,
        ip_address: ip || 'unknown',
        user_agent: userAgent || 'unknown'
      }
    })
  ];
  if (url.is_one_time) {
    transactionRequests[0] = prisma.url.update({
      where: { id: url.id },
      data: { click_count: { increment: 1 }, is_active: false }
    });
  }
  prisma.$transaction(transactionRequests).catch(err => {
    console.error('Error logging visit analytics:', err);
  });
  return {
    original_url: url.original_url,
    click_count: url.click_count + 1
  };
};

File: backend/src/main/java/com/urlshortener/service/UrlService.java
package com.urlshortener.service;

import com.urlshortener.cache.LRUCache;
import com.urlshortener.dto.request.CreateUrlRequest;
import com.urlshortener.dto.request.UpdateUrlRequest;
import com.urlshortener.entity.Url;
import com.urlshortener.entity.Visit;
import com.urlshortener.exception.ApiException;
import com.urlshortener.repository.UrlRepository;
import com.urlshortener.repository.VisitRepository;
import com.urlshortener.util.ShortCodeGenerator;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.scheduling.annotation.Async;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.*;

@Service
public class UrlService {

    @Autowired
    private UrlRepository urlRepository;

    @Autowired
    private VisitRepository visitRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private LRUCache<String, Object> urlCache;

    @Value("${client.url}")
    private String clientUrl;

    @Transactional
    public Url createShortUrl(CreateUrlRequest request, UUID userId) {
        String customAlias = request.getCustomAlias();
        Instant expiresAt = parseExpiresAt(request.getExpiresAt());
        String password = request.getPassword();
        Boolean isOneTime = request.getIsOneTime();

        if (userId == null) {
            customAlias = null;
            expiresAt = Instant.now().plus(7, ChronoUnit.DAYS);
        }

        String shortCode;

        if (customAlias != null && !customAlias.trim().isEmpty()) {
            
            Optional<Url> conflict = urlRepository.findByShortCodeOrCustomAlias(customAlias, customAlias);
            if (conflict.isPresent()) {
                throw new ApiException("This custom alias is already taken.", 400);
            }
            shortCode = customAlias;
        } else {
            
            shortCode = generateUniqueShortCode();
            customAlias = null;
        }

        String passwordHash = null;
        if (password != null && !password.trim().isEmpty()) {
            passwordHash = passwordEncoder.encode(password);
        }

        Url url = new Url();
        url.setOriginalUrl(request.getOriginalUrl());
        url.setShortCode(shortCode);
        url.setCustomAlias(customAlias);
        url.setClickCount(0);
        url.setExpiresAt(expiresAt);
        url.setPasswordHash(passwordHash);
        url.setOneTime(isOneTime != null && isOneTime);
        url.setActive(true);
        url.setUserId(userId);
        url.setCreatedAt(Instant.now());

        url = urlRepository.save(url);

        return url;
    }

    private String generateUniqueShortCode() {
        for (int i = 0; i < 10; i++) {
            String code = ShortCodeGenerator.generate(6);
            if (urlRepository.findByShortCode(code).isEmpty()) {
                return code;
            }
        }
        throw new ApiException("Failed to generate a unique short code. Please try again.", 500);
    }

    public Map<String, Object> getUrls(UUID userId, String search, Boolean isActive, int page, int limit) {
        Page<Url> urlPage = urlRepository.findByUserIdWithFilters(
                userId, search, isActive, PageRequest.of(page - 1, limit));

        Map<String, Object> pagination = new LinkedHashMap<>();
        pagination.put("total", urlPage.getTotalElements());
        pagination.put("page", page);
        pagination.put("limit", limit);
        pagination.put("pages", urlPage.getTotalPages());

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("urls", urlPage.getContent());
        result.put("pagination", pagination);

        return result;
    }

    @Transactional
    public Url updateUrl(UUID urlId, UUID userId, UpdateUrlRequest request) {
        Url url = urlRepository.findById(urlId)
                .orElseThrow(() -> new ApiException("URL not found.", 404));

        if (!url.getUserId().equals(userId)) {
            throw new ApiException("You do not have permission to modify this URL.", 403);
        }

        String oldShortCode = url.getShortCode();
        String oldCustomAlias = url.getCustomAlias();

        if (request.getOriginalUrl() != null) {
            url.setOriginalUrl(request.getOriginalUrl());
        }

        if (request.getCustomAlias() != null) {
            String newAlias = request.getCustomAlias().trim();
            if (!newAlias.isEmpty()) {
                
                if (!newAlias.equals(url.getCustomAlias())) {
                    
                    Optional<Url> conflict = urlRepository.findByShortCodeOrCustomAlias(newAlias, newAlias);
                    if (conflict.isPresent() && !conflict.get().getId().equals(url.getId())) {
                        throw new ApiException("This custom alias is already taken.", 400);
                    }
                    url.setCustomAlias(newAlias);
                    url.setShortCode(newAlias);
                }
            } else {
                
                url.setCustomAlias(null);
                url.setShortCode(generateUniqueShortCode());
            }
        }

        if (request.getExpiresAt() != null) {
            url.setExpiresAt(parseExpiresAt(request.getExpiresAt()));
        }

        if (request.getPassword() != null) {
            if (request.getPassword().trim().isEmpty()) {
                
                url.setPasswordHash(null);
            } else {
                url.setPasswordHash(passwordEncoder.encode(request.getPassword()));
            }
        }

        if (request.getIsOneTime() != null) {
            url.setOneTime(request.getIsOneTime());
        }

        if (request.getIsActive() != null) {
            url.setActive(request.getIsActive());
        }

        urlCache.remove(oldShortCode);
        if (oldCustomAlias != null) {
            urlCache.remove(oldCustomAlias);
        }
        
        if (url.getShortCode() != null && !url.getShortCode().equals(oldShortCode)) {
            urlCache.remove(url.getShortCode());
        }
        if (url.getCustomAlias() != null && !url.getCustomAlias().equals(oldCustomAlias)) {
            urlCache.remove(url.getCustomAlias());
        }

        url = urlRepository.save(url);
        return url;
    }

    @Transactional
    public void deleteUrl(UUID urlId, UUID userId) {
        Url url = urlRepository.findById(urlId)
                .orElseThrow(() -> new ApiException("URL not found.", 404));

        if (!url.getUserId().equals(userId)) {
            throw new ApiException("You do not have permission to delete this URL.", 403);
        }

        urlCache.remove(url.getShortCode());
        if (url.getCustomAlias() != null) {
            urlCache.remove(url.getCustomAlias());
        }

        urlRepository.delete(url);
    }

    @Transactional
    public Map<String, Object> resolveUrlAndTrackVisit(String shortCode, String ip, String userAgent, String password) {
        
        Url url = (Url) urlCache.get(shortCode);

        if (url == null) {
            
            url = urlRepository.findByShortCodeOrCustomAlias(shortCode, shortCode)
                    .orElseThrow(() -> new ApiException("Short URL not found.", 404));

            urlCache.put(url.getShortCode(), url);
            if (url.getCustomAlias() != null) {
                urlCache.put(url.getCustomAlias(), url);
            }
        }

        if (!url.isActive()) {
            throw new ApiException("This short URL has been deactivated.", 400);
        }

        if (url.getExpiresAt() != null && url.getExpiresAt().isBefore(Instant.now())) {
            throw new ApiException("This short URL has expired.", 410);
        }

        if (url.getPasswordHash() != null) {
            if (password == null || password.trim().isEmpty()) {
                throw new ApiException("PASSWORD_REQUIRED", 403);
            }
            if (!passwordEncoder.matches(password, url.getPasswordHash())) {
                throw new ApiException("Invalid password.", 403);
            }
        }

        int currentClickCount = url.getClickCount();
        trackVisitAsync(url, ip, userAgent);

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("original_url", url.getOriginalUrl());
        result.put("click_count", currentClickCount + 1);

        return result;
    }

    private void trackVisitAsync(Url url, String ip, String userAgent) {
        try {
            
            url.setClickCount(url.getClickCount() + 1);

            if (url.isOneTime()) {
                url.setActive(false);
                
                urlCache.remove(url.getShortCode());
                if (url.getCustomAlias() != null) {
                    urlCache.remove(url.getCustomAlias());
                }
            }

            urlRepository.save(url);

            Visit visit = new Visit();
            visit.setUrlId(url.getId());
            visit.setIpAddress(ip);
            visit.setUserAgent(userAgent);
            visit.setVisitedAt(Instant.now());
            visitRepository.save(visit);
        } catch (Exception e) {
            
            System.err.println("Error tracking visit: " + e.getMessage());
        }
    }

    private Instant parseExpiresAt(String expiresAtStr) {
        if (expiresAtStr == null || expiresAtStr.trim().isEmpty() || "null".equalsIgnoreCase(expiresAtStr.trim())) {
            return null;
        }
        try {
            return Instant.parse(expiresAtStr);
        } catch (Exception e) {
            try {
                return java.time.OffsetDateTime.parse(expiresAtStr).toInstant();
            } catch (Exception e2) {
                throw new ApiException("Invalid expiration date format.", 400);
            }
        }
    }
}
